packages/panzer/adapters-stk/test/projection/project_field.cpp

```diff
@@ -0,0 +1,676 @@
+  for (size_t i=0; i < orts_host.extent(0); ++i)
+    orts_host(i) = orts->at(i);
+  Kokkos::deep_copy(orts_dev,orts_host);
+
+  auto it2basis = basis->template getIntrepid2Basis<PHX::exec_space,double,double>();
+  auto functionSpace = it2basis->getFunctionSpace();
+  auto cell_topology = it2basis->getBaseCellTopology(); // See note above
+
+  {
+    Intrepid2::Experimental::ProjectionStruct<PHX::Device,double> projStruct;
+    projStruct.createL2ProjectionStruct(it2basis.get(), 3); // cubature order 3
+
+    int numPoints = projStruct.getNumTargetEvalPoints();
+    DynRankView evaluationPoints("evaluationPoints", numOwnedElems, numPoints, dim);
+
+    pts::getL2EvaluationPoints(evaluationPoints,
+        orts_dev,
+        it2basis.get(),
+        &projStruct);
+
+    DynRankView refTargetAtEvalPoints, physTargetAtEvalPoints;
+    if(functionSpace == Intrepid2::FUNCTION_SPACE_HCURL || functionSpace == Intrepid2::FUNCTION_SPACE_HDIV) {
+      refTargetAtEvalPoints = DynRankView("targetAtEvalPoints", numOwnedElems, numPoints, dim);
+      physTargetAtEvalPoints = DynRankView("targetAtEvalPoints", numOwnedElems, numPoints, dim);
+    } else {
+      refTargetAtEvalPoints = DynRankView("targetAtEvalPoints", numOwnedElems, numPoints);
+      physTargetAtEvalPoints = DynRankView("targetAtEvalPoints", numOwnedElems, numPoints);
+    }
+
+    auto eShape = elemShape;
+    DynRankView physEvalPoints("physEvalPoints", numOwnedElems, numPoints, dim);
+    {
+      DynRankView linearBasisValuesAtEvalPoint("linearBasisValuesAtEvalPoint", numOwnedElems, numNodesPerElem);
+
+      Kokkos::parallel_for(Kokkos::RangePolicy<typename PHX::exec_space>(0,numOwnedElems),
+          KOKKOS_LAMBDA (const int &i) {
+        auto basisValuesAtEvalPoint = Kokkos::subview(linearBasisValuesAtEvalPoint,i,Kokkos::ALL());
+        for(int j=0; j<numPoints; ++j){
+          auto evalPoint = Kokkos::subview(evaluationPoints,i,j,Kokkos::ALL());
+          switch (eShape) {
+          case HEX:
+            Intrepid2::Impl::Basis_HGRAD_HEX_C1_FEM::template Serial<Intrepid2::OPERATOR_VALUE>::getValues(basisValuesAtEvalPoint, evalPoint);
+            break;
+          case TET:
+            Intrepid2::Impl::Basis_HGRAD_TET_C1_FEM::template Serial<Intrepid2::OPERATOR_VALUE>::getValues(basisValuesAtEvalPoint, evalPoint);
+            break;
+          case QUAD:
+            Intrepid2::Impl::Basis_HGRAD_QUAD_C1_FEM::template Serial<Intrepid2::OPERATOR_VALUE>::getValues(basisValuesAtEvalPoint, evalPoint);
+            break;
+          case TRI:
+            Intrepid2::Impl::Basis_HGRAD_TRI_C1_FEM::template Serial<Intrepid2::OPERATOR_VALUE>::getValues(basisValuesAtEvalPoint, evalPoint);
+            break;
+          }
+          for(size_t k=0; k<numNodesPerElem; ++k)
+            for(int d=0; d<dim; ++d)
+              physEvalPoints(i,j,d) += cellNodes(i,k,d)*basisValuesAtEvalPoint(k);
+        }
+      });
+      Kokkos::fence();
+    }
```

Comment:
@mperego Any ideas why this might cause issues when we go to multiple worksets for this evaluator?
I get a double free error associated with `linearBasisValuesAtEvalPoint`. By way of commenting things out within this block, the error seems to be tied to the `getValues` calls (i.e., the error goes away when I comment out the `switch` block). 

FYI - I basically copied this from something you had written.

@rppawlo

Reply:
what issues do you get? I cannot see anything obviously wrong

Reply:
I see the following error:

```
*** Error in `PanzerAdaptersSTK_ProjectField.exe': double free or corruption (!prev): 0x0000000003d37dc0 ***
```
```
bt with gdb:

#0  0x00007ffff470f387 in raise () from /lib64/libc.so.6
#1  0x00007ffff4710a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff4751f67 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff475a329 in _int_free () from /lib64/libc.so.6
#4  0x00000000022edb83 in Kokkos::HostSpace::impl_deallocate (this=0x3cf9160, arg_label=0x3c15550 "linearBasisValuesAtEvalPoint", arg_alloc_ptr=0x3d02040, arg_alloc_size=416, arg_logical_size=288, arg_handle=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_HostSpace.cpp:212
#5  0x00000000022eda36 in Kokkos::HostSpace::deallocate (this=0x3cf9160, arg_label=0x3c15550 "linearBasisValuesAtEvalPoint", arg_alloc_ptr=0x3d02040, arg_alloc_size=416, arg_logical_size=288) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_HostSpace.cpp:196
#6  0x00000000022edc49 in Kokkos::Impl::SharedAllocationRecord<Kokkos::HostSpace, void>::~SharedAllocationRecord (this=0x3cf9100, __in_chrg=<optimized out>) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_HostSpace.cpp:236
#7  0x0000000000587512 in Kokkos::Impl::SharedAllocationRecord<Kokkos::HostSpace, Kokkos::Impl::ViewValueFunctor<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, double, true> >::~SharedAllocationRecord (this=0x3cf9100, __in_chrg=<optimized out>)
    at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:281
#8  0x000000000058752e in Kokkos::Impl::SharedAllocationRecord<Kokkos::HostSpace, Kokkos::Impl::ViewValueFunctor<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, double, true> >::~SharedAllocationRecord (this=0x3cf9100, __in_chrg=<optimized out>)
    at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:281
#9  0x000000000041c161 in Kokkos::Impl::(anonymous namespace)::deallocate<Kokkos::HostSpace, Kokkos::Impl::ViewValueFunctor<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, double, true> > (record_ptr=0x3cf9100) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:263
#10 0x00000000022f5554 in Kokkos::Impl::SharedAllocationRecord<void, void>::decrement (arg_record=0x3cf9100) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_SharedAlloc.cpp:265
#11 0x000000000043f726 in ~SharedAllocationTracker (this=0x7ffffffea3e0, __in_chrg=<optimized out>) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:419
#12 Kokkos::DynRankView<double, Kokkos::Serial>::~DynRankView (this=0x7ffffffea3e0, __in_chrg=<optimized out>) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/kokkos/containers/src/Kokkos_DynRankView.hpp:968
#13 0x00000000006a763b in panzer_stk::GetCoeffsEvaluator<panzer::Traits::Residual, panzer::Traits>::evaluateFields (this=0x3ceef50, workset=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/panzer/adapters-stk/test/projection/project_field.cpp:297
#14 0x00000000006bf79f in PHX::DagManager<panzer::Traits>::evaluateFields (this=0x3cedbb8, d=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/phalanx/src/Phalanx_DAG_Manager_Def.hpp:494
#15 0x000000000069ce9f in PHX::EvaluationContainer<panzer::Traits::Residual, panzer::Traits>::evaluateFields (this=0x3cedbb0, d=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/phalanx/src/Phalanx_EvaluationContainer_Def.hpp:455
#16 0x000000000046a6e2 in PHX::FieldManager<panzer::Traits>::evaluateFields<panzer::Traits::Residual> (this=0x3ce5960, d=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/phalanx/src/Phalanx_FieldManager_Def.hpp:312
#17 0x000000000043843b in checkProjection<panzer::Traits::Residual> (mesh=..., fmap=..., eShape=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/panzer/adapters-stk/test/projection/project_field.cpp:624
#18 0x0000000000425eb4 in project_field_value_UnitTest<panzer::Traits::Residual>::runUnitTestImpl (this=0x3a827a0 <instance_project_field_ResidualType_value_UnitTest>, out=..., success=@0x7fffffff44cf: true) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/panzer/adapters-stk/test/projection/project_field.cpp:450
#19 0x00000000022a8850 in Teuchos::UnitTestBase::runUnitTest (this=0x3a827a0 <instance_project_field_ResidualType_value_UnitTest>, out=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestBase.cpp:62
#20 0x00000000022ada99 in Teuchos::UnitTestRepository::runUnitTestImpl (unitTest=..., out=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestRepository.cpp:539
#21 0x00000000022ac40c in Teuchos::UnitTestRepository::runUnitTests (out=...) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestRepository.cpp:317
#22 0x00000000022ad481 in Teuchos::UnitTestRepository::runUnitTestsFromMain (argc=1, argv=0x7fffffff52e8) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/teuchos/core/src/Teuchos_UnitTestRepository.cpp:423
#23 0x0000000000ae36c0 in main (argc=1, argv=0x7fffffff52e8) at /ascldap/users/bwreute/hpc_home/Software/myTrilinosFork/Trilinos/packages/phalanx/test/Utilities/Phalanx_UnitTestMain.cpp:19
```

The `getValues` calls are very simple but this error goes away when commenting them out.

Reply:
I would check that `numNodesPerElem` is correct (3 for Triangles,  4 for Quads and Tets, 8 for Hexas). 
Also, if you build trilinos in debug mode, maybe you'll find out something.